File: android/app/src/main/java/com/dibay/app/nativevoice/NativeVoiceCallUiPresenter.java
package com.dibay.app.nativevoice;

import android.content.Context;
import com.dibay.app.IncomingCallUiCopy;
import com.dibay.app.R;

/** Maps Native Voice Runtime session/state to render-only UI model. No ownership of session/Agora/cleanup. */
public final class NativeVoiceCallUiPresenter {
  public enum Phase {
    INCOMING,
    DIALING,
    CONNECTING,
    CONNECTED,
    ENDING
  }

  public static final class Model {
    public final Phase phase;
    public final String peerName;
    public final String statusText;
    public final String avatarInitial;
    public final boolean showIncomingActions;
    public final boolean showMediaActions;
    public final boolean micChromeEnabled;
    public final boolean showDuration;

    Model(
        Phase phase,
        String peerName,
        String statusText,
        String avatarInitial,
        boolean showIncomingActions,
        boolean showMediaActions,
        boolean micChromeEnabled,
        boolean showDuration) {
      this.phase = phase;
      this.peerName = peerName;
      this.statusText = statusText;
      this.avatarInitial = avatarInitial;
      this.showIncomingActions = showIncomingActions;
      this.showMediaActions = showMediaActions;
      this.micChromeEnabled = micChromeEnabled;
      this.showDuration = showDuration;
    }
  }

  private NativeVoiceCallUiPresenter() {}

  public static Model build(Context context, NativeVoiceCallRuntime.Session session, NativeVoiceCallRuntime.State state) {
    Context app = context != null ? context.getApplicationContext() : null;
    String peerName = resolvePeerName(session);
    Phase phase = resolvePhase(session, state);
    String statusText = resolveStatusText(app, phase);
    String avatarInitial = IncomingCallUiCopy.peerInitial(peerName);
    boolean ending = phase == Phase.ENDING;
    boolean incoming = phase == Phase.INCOMING;
    boolean connected = phase == Phase.CONNECTED;
    boolean dialingOrConnecting = phase == Phase.DIALING || phase == Phase.CONNECTING;
    return new Model(
        phase,
        peerName,
        statusText,
        avatarInitial,
        incoming && !ending,
        (dialingOrConnecting || connected) && !ending,
        connected && !ending,
        connected && !ending);
  }

  public static Phase resolvePhase(NativeVoiceCallRuntime.Session session, NativeVoiceCallRuntime.State state) {
    if (state == NativeVoiceCallRuntime.State.ENDING
        || state == NativeVoiceCallRuntime.State.ENDED
        || state == NativeVoiceCallRuntime.State.FAILED) {
      return Phase.ENDING;
    }
    if (state == NativeVoiceCallRuntime.State.CONNECTED) {
      return Phase.CONNECTED;
    }
    if (state == NativeVoiceCallRuntime.State.RINGING) {
      return Phase.INCOMING;
    }
    if (session != null && session.initiator && state == NativeVoiceCallRuntime.State.CONNECTING) {
      return Phase.DIALING;
    }
    if (state == NativeVoiceCallRuntime.State.ACCEPTING || state == NativeVoiceCallRuntime.State.CONNECTING) {
      return Phase.CONNECTING;
    }
    return Phase.CONNECTING;
  }

  private static String resolvePeerName(NativeVoiceCallRuntime.Session session) {
    if (session == null || session.callerName == null || session.callerName.trim().isEmpty()) {
      return "DIBAY";
    }
    return IncomingCallUiCopy.sanitizeNickname(session.callerName.trim());
  }

  private static String resolveStatusText(Context app, Phase phase) {
    if (app == null) return "";
    switch (phase) {
      case INCOMING:
        return safeString(app, R.string.dibay_voice_call_incoming);
      case DIALING:
        return safeString(app, R.string.dibay_voice_call_dialing);
      case CONNECTING:
        return safeString(app, R.string.dibay_voice_call_connecting);
      case CONNECTED:
        return safeString(app, R.string.dibay_voice_call_connected);
      case ENDING:
      default:
        return safeString(app, R.string.dibay_voice_call_ending);
    }
  }

  private static String safeString(Context app, int resId) {
    try {
      return app.getString(resId);
    } catch (RuntimeException error) {
      return "";
    }
  }
}

File: ios/App/App/Call/NativeVoiceCallUiPresenter.swift
import Foundation

/** Maps Native Voice Runtime session/phase to render-only UI model (Android `NativeVoiceCallUiPresenter` parity). */
enum NativeVoiceCallUiPresenter {
  enum Phase: Equatable, Sendable {
    case incoming
    case dialing
    case connecting
    case connected
    case ending
  }

  struct Model: Equatable, Sendable {
    let phase: Phase
    let peerName: String
    let statusText: String
    let avatarInitial: String
    let showIncomingActions: Bool
    let showMediaActions: Bool
    let micChromeEnabled: Bool
    let showDuration: Bool
  }

  static func build(session: NativeVoiceCallSession?, phase: NativeVoiceCallPhase) -> Model {
    let peerName = resolvePeerName(session)
    let uiPhase = resolvePhase(session: session, phase: phase)
    let statusText = resolveStatusText(uiPhase)
    let avatarInitial = initialFromName(peerName)
    let ending = uiPhase == .ending
    let incoming = uiPhase == .incoming
    let connected = uiPhase == .connected
    let dialingOrConnecting = uiPhase == .dialing || uiPhase == .connecting
    return Model(
      phase: uiPhase,
      peerName: peerName,
      statusText: statusText,
      avatarInitial: avatarInitial,
      showIncomingActions: incoming && !ending,
      showMediaActions: (dialingOrConnecting || connected) && !ending,
      micChromeEnabled: connected && !ending,
      showDuration: connected && !ending
    )
  }

  static func resolvePhase(session: NativeVoiceCallSession?, phase: NativeVoiceCallPhase) -> Phase {
    switch phase {
    case .ending, .ended, .failed, .rejecting, .idle:
      return .ending
    case .connected:
      return .connected
    case .incomingPresented:
      return .incoming
    case .outgoingStarting:
      return .dialing
    case .joining:
      return .connecting
    case .accepting, .accepted, .tokenPending:
      return .connecting
    }
  }

  private static func resolvePeerName(_ session: NativeVoiceCallSession?) -> String {
    let name = session?.callerName.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    if name.isEmpty { return "DIBAY" }
    return sanitizeNickname(name)
  }

  private static func sanitizeNickname(_ value: String) -> String {
    let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
    if trimmed.hasPrefix("@") { return String(trimmed.dropFirst()) }
    return trimmed
  }

  private static func initialFromName(_ peerName: String) -> String {
    let trimmed = peerName.trimmingCharacters(in: .whitespacesAndNewlines)
    guard let first = trimmed.first else { return "D" }
    return String(first).uppercased()
  }

  private static func resolveStatusText(_ phase: Phase) -> String {
    switch phase {
    case .incoming:
      return Copy.incoming
    case .dialing:
      return Copy.dialing
    case .connecting:
      return Copy.connecting
    case .connected:
      return Copy.connected
    case .ending:
      return Copy.ending
    }
  }

  private enum Copy {
    static let incoming = "음성 통화 수신"
    static let dialing = "전화 거는 중 …"
    static let connecting = "연결 중…"
    static let connected = "통화 중"
    static let ending = "통화 종료 중"
  }
}
